spec/ics-18-relayer-algorithms/README.md

```diff
@@ -0,0 +1,69 @@
+---
+ics: 18
+title: Relayer Algorithms
+stage: draft
+category: ibc-misc
+author: Christopher Goes <cwgoes@tendermint.com>
+created: 2019-03-07
+modified: 2019-04-02
+---
+
+## Synopsis
+
+Relayer algorithms are the "physical" connection layer of IBC — off-chain processes responsible for relaying data between two chains running the IBC protocol by scanning the state of each chain, constructing appropriate datagrams, and executing them on the opposite chain as allowed by the protocol.
+
+## Specification
+
+### Motivation
+
+In the IBC protocol, a blockchain can only record the *intention* to send particular data to another chain. Physical datagram relay must be performed by off-chain infrastructure. This standard defines the concept of a *relayer* algorithm, executable by an off-chain process with the ability to query chain state, to perform this relay.
+
+### Desired Properties
+
+- No safety properties of IBC should depend on relayer behavor (assume Byzantine relayers).
+- Liveness properties of IBC should depend only on the existence of at least one correct, live relayer.
+- Relaying should be permissionless, all requisite verification should be performed on-chain.
+- Requisite communication between the IBC user and the relayer should be minimized.
+- Provision for relayer incentivization should be possible at the application layer.
+
+### Technical Specification
+
+#### Relayer Algorithm
+
+The relayer algorithm is defined over a set `C` of chains implementing the IBC protocol.
+
+`pendingDatagrams` calculates the set of all valid datagrams to be relayed from one chain to another based on the state of both chains. Subcomponents of this function are defined in individual ICSs. The relayer must possess prior knowledge of what subset of the IBC protocol is implemented by the blockchains in the set for which they are relaying (e.g. by reading the source code).
+
+`submitDatagram` is a procedure defined per-chain (submitting a transaction of some sort).
+
+![Relayer](relayer.png)
+
+#### Incentivization
+
+The relay process must have access to accounts on both chains with sufficient balance to pay for transaction fees. Relayers may employ application-level methods to recoup these fees.
+
+Any number of relayer processes may be safely run in parallel (and indeed, it is expected that separate relayers will serve separate subsets of the interchain). However, they may consume unnecessary fees if they submit the same proof multiple times, so some minimal coordination may be ideal.
```

Comment:
> consume unnecessary fees if they submit the same proof multiple times,

What if a relayer sends the same proof 1000 times?

> so some minimal coordination may be ideal.

does this mean that for someone who's initiating IBC transaction, he/she should have enough fees for all existing and future relayers?? Wouldn't it make more sense to just give fees to the first relayer?

Reply:
> What if a relayer sends the same proof 1000 times?

Nothing (it's just like sending a transaction which tries to spend the same funds 1000 times, the first transaction will spend the funds and the remaining 999 will fail).

> does this mean that for someone who's initiating IBC transaction, he/she should have enough fees for all existing and future relayers?? Wouldn't it make more sense to just give fees to the first relayer?

Hmm, no, this has nothing to do with what a user of IBC would do - they would just include a small payment for whichever relayer submits their packet.